### ISSUE TITLE
Potential bug when using a unary minus

### ISSUE THREAD

**Issue by ludkinm** [autodiff/autodiff#41]
I've spotted the following bug when using the dual type in ``forward.hpp``
```c++
dual f_bug(dual x)
{
  return -x - 2.0 * log(1.0 + exp(-x));
}
```
Running ``f(1.0)`` should result in ``-1.62652...`` however I get ``-1.0``

Adding parentheses fixes the problem:
```c++
dual f_ok(dual x)
{
  return -(x + 2.0 * log(1.0 + exp(-x)));
}
```
Is there an issue with the order of operations? Or is this expected behaviour?

**Comment by allanleal** [comment]
I'm checking here.

**Comment by allanleal** [comment]
Yes, I can confirm there is a bug - a difficult one to explain, since it is very subtle and it involves local variables (`BinaryExpr<L,R>`) being taken as reference inside an outer expression node (`NegExpr<BinaryExpr<L,R>&>` instead of `NegExpr<BinaryExpr<L,R>>`) because I was relying on automatic return type deduction in some places! A fix is on the way.

**Comment by allanleal** [comment]
Just waiting for CI to pass to create a new release. Thanks for reporting this!

**Comment by ludkinm** [comment]
No problem, thanks for the fix.

**Comment by allanleal** [comment]
Fixed on release v0.5.5.
